File: src/apps/bench/rustbenchs/src/bsyscall.rs
/*
 * Copyright (C) 2018 Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * Copyright (C) 2019-2022 Nils Asmussen, Barkhausen Institut
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

use m3::cap::SelSpace;
use m3::cell::StaticCell;
use m3::cfg;
use m3::com::{EpMng, MemCap, MemGate, Perm, RecvCap, RecvGate};
use m3::kif;
use m3::mem::{GlobOff, VirtAddr};
use m3::rc::Rc;
use m3::syscalls;
use m3::test::WvTester;
use m3::tiles::{Activity, ActivityArgs, ChildActivity, Tile};
use m3::time::{CycleInstant, Profiler, Runner};
use m3::util::math;
use m3::{println, wv_assert_ok, wv_perf, wv_run_test};

static SEL: StaticCell<kif::CapSel> = StaticCell::new(0);

pub fn run(t: &mut dyn WvTester) {
    SEL.set(SelSpace::get().alloc_sel());

    wv_run_test!(t, noop);
    wv_run_test!(t, activate);
    wv_run_test!(t, create_mgate);
    wv_run_test!(t, create_rgate);
    wv_run_test!(t, create_sgate);
    wv_run_test!(t, create_map);
    wv_run_test!(t, create_srv);
    wv_run_test!(t, derive_mem);
    wv_run_test!(t, exchange);
    wv_run_test!(t, revoke_mem_gate);
    wv_run_test!(t, revoke_recv_gate);
    wv_run_test!(t, revoke_send_gate);
}

fn noop(_t: &mut dyn WvTester) {
    let prof = Profiler::default();

    wv_perf!(
        "noop",
        prof.run::<CycleInstant, _>(|| {
            wv_assert_ok!(syscalls::noop());
        })
    );
}

fn activate(_t: &mut dyn WvTester) {
    let mcap = wv_assert_ok!(MemCap::new(0x1000, Perm::RW));
    let ep = wv_assert_ok!(EpMng::get().acquire(0));

    let prof = Profiler::default();

    wv_perf!(
        "activate",
        prof.run::<CycleInstant, _>(|| {
            wv_assert_ok!(syscalls::activate(
                ep.sel(),
                mcap.sel(),
                kif::INVALID_SEL,
                0
            ));
        })
    );

    EpMng::get().release(ep, true);
}

fn create_mgate(_t: &mut dyn WvTester) {
    let prof = Profiler::default().repeats(100).warmup(100);

    #[derive(Default)]
    struct Tester(VirtAddr);

    impl Runner for Tester {
        fn run(&mut self) {
            wv_assert_ok!(syscalls::create_mgate(
                SEL.get(),
                Activity::own().sel(),
                self.0,
                cfg::PAGE_SIZE as GlobOff,
                Perm::R
            ));
        }

        fn post(&mut self) {
            wv_assert_ok!(syscalls::revoke(
                Activity::own().sel(),
                kif::CapRngDesc::new(kif::CapType::Object, SEL.get(), 1),
                true
            ));
        }
    }

    let addr = VirtAddr::from(math::round_dn(
        &create_mgate as *const _ as usize,
        cfg::PAGE_SIZE,
    ));
    wv_perf!(
        "create_mgate",
        prof.runner::<CycleInstant, _>(&mut Tester(addr))
    );
}

fn create_rgate(_t: &mut dyn WvTester) {
    let prof = Profiler::default().repeats(100).warmup(100);

    #[derive(Default)]
    struct Tester();

    impl Runner for Tester {
        fn run(&mut self) {
            wv_assert_ok!(syscalls::create_rgate(SEL.get(), 10, 10));
        }

        fn post(&mut self) {
            wv_assert_ok!(syscalls::revoke(
                Activity::own().sel(),
                kif::CapRngDesc::new(kif::CapType::Object, SEL.get(), 1),
                true
            ));
        }
    }

    wv_perf!(
        "create_rgate",
        prof.runner::<CycleInstant, _>(&mut Tester::default())
    );
}

fn create_sgate(_t: &mut dyn WvTester) {
    let prof = Profiler::default().repeats(100).warmup(10);

    #[derive(Default)]
    struct Tester(Option<RecvGate>);

    impl Runner for Tester {
        fn pre(&mut self) {
            if self.0.is_none() {
                self.0 = Some(wv_assert_ok!(RecvGate::new(10, 10)));
            }
        }

        fn run(&mut self) {
            wv_assert_ok!(syscalls::create_sgate(
                SEL.get(),
                self.0.as_ref().unwrap().sel(),
                0x1234,
                1024
            ));
        }

        fn post(&mut self) {
            wv_assert_ok!(syscalls::revoke(
                Activity::own().sel(),
                kif::CapRngDesc::new(kif::CapType::Object, SEL.get(), 1),
                true
            ));
        }
    }

    wv_perf!(
        "create_sgate",
        prof.runner::<CycleInstant, _>(&mut Tester::default())
    );
}

fn create_map(_t: &mut dyn WvTester) {
    if !Activity::own().tile_desc().has_virtmem() {
        println!("Tile has no virtual memory support; skipping");
        return;
    }

    const DEST: VirtAddr = VirtAddr::new(0x3000_0000);
    let prof = Profiler::default().repeats(100).warmup(10);

    struct Tester(MemGate);

    impl Runner for Tester {
        fn pre(&mut self) {
            // one warmup run, because the revoke leads to an unmap, which flushes and invalidates
            // all cache lines
            wv_assert_ok!(syscalls::create_map(
                DEST,
                Activity::own().sel(),
                self.0.sel(),
                0,
                1,
                Perm::RW
            ));
        }

        fn run(&mut self) {
            wv_assert_ok!(syscalls::create_map(
                DEST + cfg::PAGE_SIZE,
                Activity::own().sel(),
                self.0.sel(),
                1,
                1,
                Perm::RW
            ));
        }

        fn post(&mut self) {
            wv_assert_ok!(syscalls::revoke(
                Activity::own().sel(),
                kif::CapRngDesc::new(
                    kif::CapType::Mapping,
                    DEST.as_goff() / cfg::PAGE_SIZE as GlobOff,
                    2
                ),
                true
            ));
        }
    }

    let mut tester = Tester(MemGate::new((cfg::PAGE_SIZE * 2) as GlobOff, Perm::RW).unwrap());
    wv_perf!("create_map", prof.runner::<CycleInstant, _>(&mut tester));
}

fn create_srv(_t: &mut dyn WvTester) {
    let prof = Profiler::default().repeats(100).warmup(10);

    #[derive(Default)]
    struct Tester(Option<RecvGate>);

    impl Runner for Tester {
        fn pre(&mut self) {
            if self.0.is_none() {
                self.0 = Some(wv_assert_ok!(RecvGate::new(10, 10)));
            }
        }

        fn run(&mut self) {
            wv_assert_ok!(syscalls::create_srv(
                SEL.get(),
                self.0.as_ref().unwrap().sel(),
                "test",
                0
            ));
        }

        fn post(&mut self) {
            wv_assert_ok!(syscalls::revoke(
                Activity::own().sel(),
                kif::CapRngDesc::new(kif::CapType::Object, SEL.get(), 1),
                true
            ));
        }
    }

    wv_perf!(
        "create_srv",
        prof.runner::<CycleInstant, _>(&mut Tester::default())
    );
}

fn derive_mem(_t: &mut dyn WvTester) {
    let prof = Profiler::default().repeats(100).warmup(10);

    #[derive(Default)]
    struct Tester(Option<MemGate>);

    impl Runner for Tester {
        fn pre(&mut self) {
            if self.0.is_none() {
                self.0 = Some(wv_assert_ok!(MemGate::new(0x1000, Perm::RW)));
            }
        }

        fn run(&mut self) {
            wv_assert_ok!(syscalls::derive_mem(
                Activity::own().sel(),
                SEL.get(),
                self.0.as_ref().unwrap().sel(),
                0,
                0x1000,
                Perm::RW
            ));
        }

        fn post(&mut self) {
            wv_assert_ok!(syscalls::revoke(
                Activity::own().sel(),
                kif::CapRngDesc::new(kif::CapType::Object, SEL.get(), 1),
                true
            ));
        }
    }

    wv_perf!(
        "derive_mem",
        prof.runner::<CycleInstant, _>(&mut Tester::default())
    );
}

fn exchange(_t: &mut dyn WvTester) {
    let prof = Profiler::default().repeats(100).warmup(10);

    struct Tester {
        act: Option<ChildActivity>,
        tile: Rc<Tile>,
    }

    impl Runner for Tester {
        fn pre(&mut self) {
            if self.act.is_none() {
                self.act = Some(wv_assert_ok!(ChildActivity::new_with(
                    self.tile.clone(),
                    ActivityArgs::new("test")
                )));
            }
        }

        fn run(&mut self) {
            wv_assert_ok!(syscalls::exchange(
                self.act.as_ref().unwrap().sel(),
                kif::CapRngDesc::new(kif::CapType::Object, kif::SEL_ACT, 1),
                SEL.get(),
                false,
            ));
        }

        fn post(&mut self) {
            wv_assert_ok!(syscalls::revoke(
                self.act.as_ref().unwrap().sel(),
                kif::CapRngDesc::new(kif::CapType::Object, SEL.get(), 1),
                true
            ));
        }
    }

    wv_perf!(
        "exchange",
        prof.runner::<CycleInstant, _>(&mut Tester {
            act: None,
            tile: wv_assert_ok!(Tile::get("compat|own")),
        })
    );
}

fn revoke_mem_gate(_t: &mut dyn WvTester) {
    let prof = Profiler::default().repeats(100).warmup(10);

    let mcap = wv_assert_ok!(MemCap::new(0x1000, Perm::RW));

    struct Tester {
        mcap: MemCap,
        _derived: Option<MemCap>,
    }

    impl Runner for Tester {
        fn pre(&mut self) {
            self._derived = Some(wv_assert_ok!(self.mcap.derive(0, 0x1000, Perm::RW)));
        }

        fn run(&mut self) {
            self._derived = None;
        }
    }

    let mut tester = Tester {
        mcap,
        _derived: None,
    };
    wv_perf!(
        "revoke_mem_gate",
        prof.runner::<CycleInstant, _>(&mut tester)
    );
}

fn revoke_recv_gate(_t: &mut dyn WvTester) {
    let prof = Profiler::default().repeats(100).warmup(10);

    #[derive(Default)]
    struct Tester();

    impl Runner for Tester {
        fn pre(&mut self) {
            wv_assert_ok!(syscalls::create_rgate(SEL.get(), 10, 10));
        }

        fn run(&mut self) {
            wv_assert_ok!(syscalls::revoke(
                Activity::own().sel(),
                kif::CapRngDesc::new(kif::CapType::Object, SEL.get(), 1),
                true
            ));
        }
    }

    wv_perf!(
        "revoke_recv_gate",
        prof.runner::<CycleInstant, _>(&mut Tester::default())
    );
}

fn revoke_send_gate(_t: &mut dyn WvTester) {
    let prof = Profiler::default().repeats(100).warmup(10);

    #[derive(Default)]
    struct Tester(Option<RecvCap>);

    impl Runner for Tester {
        fn pre(&mut self) {
            self.0 = Some(wv_assert_ok!(RecvCap::new(10, 10)));
            wv_assert_ok!(syscalls::create_sgate(
                SEL.get(),
                self.0.as_ref().unwrap().sel(),
                0x1234,
                1024
            ));
        }

        fn run(&mut self) {
            wv_assert_ok!(syscalls::revoke(
                Activity::own().sel(),
                kif::CapRngDesc::new(kif::CapType::Object, SEL.get(), 1),
                true
            ));
        }
    }

    wv_perf!(
        "revoke_send_gate",
        prof.runner::<CycleInstant, _>(&mut Tester::default())
    );
}

File: src/apps/bench/cppbenchs/benchs/bsyscall.cc
/*
 * Copyright (C) 2018 Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * Copyright (C) 2019-2022 Nils Asmussen, Barkhausen Institut
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <base/Common.h>
#include <base/KIF.h>
#include <base/Panic.h>
#include <base/col/SList.h>
#include <base/time/Profile.h>

#include <m3/Syscalls.h>
#include <m3/Test.h>
#include <m3/tiles/ChildActivity.h>

#include "../cppbenchs.h"

using namespace m3;

static capsel_t selector = ObjCap::INVALID;

NOINLINE static void noop() {
    Profile pr;
    WVPERF(__func__, pr.run<CycleInstant>([] {
        Syscalls::noop();
    }));
}

NOINLINE static void activate() {
    MemGate mgate = MemGate::create_global(0x1000, MemGate::RW);
    const EP &ep = mgate.activate();

    Profile pr;
    WVPERF(__func__, pr.run<CycleInstant>([&ep, &mgate] {
        Syscalls::activate(ep.sel(), mgate.sel(), KIF::INV_SEL, 0);
    }));
}

NOINLINE static void create_mgate() {
    static uintptr_t addr = Math::round_dn(reinterpret_cast<uintptr_t>(&create_mgate),
                                           static_cast<uintptr_t>(PAGE_SIZE));

    struct SyscallMGateRunner : public Runner {
        void run() override {
            Syscalls::create_mgate(selector, Activity::own().sel(), addr, PAGE_SIZE, KIF::Perm::R);
        }
        void post() override {
            Syscalls::revoke(Activity::own().sel(),
                             KIF::CapRngDesc(KIF::CapRngDesc::OBJ, selector, 1), true);
        }
    };

    Profile pr;
    SyscallMGateRunner runner;
    WVPERF(__func__, pr.runner<CycleInstant>(runner));
}

NOINLINE static void create_rgate() {
    struct SyscallRGateRunner : public Runner {
        void run() override {
            Syscalls::create_rgate(selector, 10, 10);
        }
        void post() override {
            Syscalls::revoke(Activity::own().sel(),
                             KIF::CapRngDesc(KIF::CapRngDesc::OBJ, selector, 1), true);
        }
    };

    Profile pr;
    SyscallRGateRunner runner;
    WVPERF(__func__, pr.runner<CycleInstant>(runner));
}

NOINLINE static void create_sgate() {
    struct SyscallSGateRunner : public Runner {
        explicit SyscallSGateRunner() : rgate(RecvGate::create(10, 10)) {
        }
        void run() override {
            Syscalls::create_sgate(selector, rgate.sel(), 0x1234, 1024);
        }
        void post() override {
            Syscalls::revoke(Activity::own().sel(),
                             KIF::CapRngDesc(KIF::CapRngDesc::OBJ, selector, 1), true);
        }

        RecvGate rgate;
    };

    Profile pr;
    SyscallSGateRunner runner;
    WVPERF(__func__, pr.runner<CycleInstant>(runner));
}

NOINLINE static void create_map() {
    if(!Activity::own().tile_desc().has_virtmem()) {
        println("Tile has no virtual memory support; skipping"_cf);
        return;
    }

    constexpr capsel_t DEST = 0x3000'0000 >> PAGE_BITS;

    struct SyscallMapRunner : public Runner {
        explicit SyscallMapRunner() : mgate(MemGate::create_global(PAGE_SIZE * 2, MemGate::RW)) {
        }

        void pre() override {
            // one warmup run, because the revoke leads to an unmap, which flushes and invalidates
            // all cache lines
            Syscalls::create_map(DEST, Activity::own().sel(), mgate.sel(), 0, 1, MemGate::RW);
        }

        void run() override {
            Syscalls::create_map(DEST + 1, Activity::own().sel(), mgate.sel(), 1, 1, MemGate::RW);
        }
        void post() override {
            Syscalls::revoke(Activity::own().sel(), KIF::CapRngDesc(KIF::CapRngDesc::MAP, DEST, 2),
                             true);
        }

        MemGate mgate;
    };

    Profile pr;
    SyscallMapRunner runner;
    WVPERF(__func__, pr.runner<CycleInstant>(runner));
}

NOINLINE static void create_srv() {
    struct SyscallSrvRunner : public Runner {
        explicit SyscallSrvRunner() : rgate(RecvGate::create(10, 10)) {
            rgate.activate();
        }

        void run() override {
            Syscalls::create_srv(selector, rgate.sel(), "test", 0);
        }
        void post() override {
            Syscalls::revoke(Activity::own().sel(),
                             KIF::CapRngDesc(KIF::CapRngDesc::OBJ, selector, 1), true);
        }

        RecvGate rgate;
    };

    Profile pr;
    SyscallSrvRunner runner;
    WVPERF(__func__, pr.runner<CycleInstant>(runner));
}

NOINLINE static void derive_mem() {
    struct SyscallDeriveRunner : public Runner {
        explicit SyscallDeriveRunner() : mgate(MemGate::create_global(0x1000, MemGate::RW)) {
        }

        void run() override {
            Syscalls::derive_mem(Activity::own().sel(), selector, mgate.sel(), 0, 0x1000,
                                 MemGate::RW);
        }
        void post() override {
            Syscalls::revoke(Activity::own().sel(),
                             KIF::CapRngDesc(KIF::CapRngDesc::OBJ, selector, 1), true);
        }

        MemGate mgate;
    };

    Profile pr;
    SyscallDeriveRunner runner;
    WVPERF(__func__, pr.runner<CycleInstant>(runner));
}

NOINLINE static void exchange() {
    struct SyscallExchangeRunner : public Runner {
        explicit SyscallExchangeRunner() : tile(Tile::get("own|core")), act(tile, "test") {
        }

        void run() override {
            Syscalls::exchange(act.sel(), KIF::CapRngDesc(KIF::CapRngDesc::OBJ, KIF::SEL_ACT, 1),
                               selector, false);
        }
        void post() override {
            Syscalls::revoke(act.sel(), KIF::CapRngDesc(KIF::CapRngDesc::OBJ, selector, 1), true);
        }

        Reference<Tile> tile;
        ChildActivity act;
    };

    Profile pr;
    SyscallExchangeRunner runner;
    WVPERF(__func__, pr.runner<CycleInstant>(runner));
}

NOINLINE static void revoke() {
    struct SyscallRevokeRunner : public Runner {
        void pre() override {
            mgate = new MemGate(MemGate::create_global(0x1000, MemGate::RW));
        }
        void run() override {
            delete mgate;
            mgate = nullptr;
        }

        MemGate *mgate;
    };

    Profile pr;
    SyscallRevokeRunner runner;
    WVPERF(__func__, pr.runner<CycleInstant>(runner));
}

void bsyscall() {
    selector = SelSpace::get().alloc_sel();

    RUN_BENCH(noop);
    RUN_BENCH(activate);
    RUN_BENCH(create_mgate);
    RUN_BENCH(create_rgate);
    RUN_BENCH(create_sgate);
    RUN_BENCH(create_map);
    RUN_BENCH(create_srv);
    RUN_BENCH(derive_mem);
    RUN_BENCH(exchange);
    RUN_BENCH(revoke);
}
